projects/wizard_of_wikipedia/generator/agents.py

```diff
@@ -0,0 +1,299 @@
+}
+
+
+class _GenericWizardAgent(TransformerGeneratorAgent):
+    # todo:
+    # - dict agent stuff
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        argparser.set_defaults(**DEFAULT_OPTS)
+        super(_GenericWizardAgent, cls).add_cmdline_args(argparser)
+
+    def batchify(self, obs_batch):
+        batch = super().batchify(obs_batch)
+        reordered_observations = [obs_batch[i] for i in batch.valid_indices]
+
+        checked_sentences = []
+        for obs in reordered_observations:
+            checked_sentence = '{} {} {}'.format(
+                obs.get('title', ''), TOKEN_KNOWLEDGE, obs.get('checked_sentence', '')
+            )
+            checked_sentences.append(checked_sentence)
+
+        batch['checked_sentence'] = checked_sentences
+        return batch
+
+
+class TwoStageAgent(_GenericWizardAgent):
+    def observe(self, obs):
+        if 'text' not in obs:
+            return obs
+
+        # TODO: resolve this with #1421
+        # get the dialog stuff
+        reply = self.last_reply()
+        self.observation = self.get_dialog_history(obs, reply=reply)
+        # we need to store the old text so that we can restore it
+        oldtext = obs['text']
+
+        # now we want to force prepend the knowlege stuff
+        fields = []
+        if 'chosen_topic' in obs:
+            fields += [obs['title']]
+        if 'checked_sentence' in obs:
+            fields += [TOKEN_KNOWLEDGE, obs['checked_sentence']]
+        if obs['text'] != '':
+            fields += [TOKEN_DIALOG, obs['text']]
+        obs['text'] = ' '.join(fields)
+
+        # now vectorize with the extra knowledge. It'll all get stored in the
+        # text_vec operation, etc
+        self.vectorize(
+            obs,
+            text_truncate=self.text_truncate,
+            label_truncate=self.label_truncate
+        )
+
+        # finally we need to return the old text to the way it was
+        obs['text'] = oldtext
+        assert obs is self.observation
+
+        return obs
+
+
+class EndToEndAgent(_GenericWizardAgent):
+    def _init_cuda_buffer(self, *args, **kwargs):
+        # just do nothing
+        pass
+
+    def build_criterion(self):
+        self.criterion = EndToEndCriterion(self.opt)
+        if self.use_cuda:
+            self.criterion.cuda()
+
+    def _parse_knowledge(self, obs):
+        if 'knowledge_parsed' in obs:
+            # make a copy of the list to prevent the future padding step from
+            # being destructive
+            return list(obs['knowledge_parsed'])
+
+        checked_sentence = '{} {} {}'.format(
+            obs['title'], TOKEN_KNOWLEDGE, obs['checked_sentence']
+        )
+        # grab all the nonempty knowledge
+        obs_know = [k.strip() for k in obs.get('knowledge', '').split('\n')]
+        obs_know = [k for k in obs_know if k]
+
+        # we want the correct knowledge to always be in index 0
```

Comment:
I'm a little lost on this part. Why do you force it to be first?

Reply:
IIRC later on I use that fact to make loss computation easier or something.